pandas/tests/test_algos.py

```diff
@@ -235,6 +236,28 @@ def test_on_index_object(self):
 
         tm.assert_almost_equal(result, expected)
 
+    def test_datetime64_dtype_array_returned(self):
+        # GH 9431
+        arr = np.array(['2015-01-03T00:00:00.000000000+0000',
+                        '2015-01-01T00:00:00.000000000+0000',
+                        '2015-01-01T00:00:00.000000000+0000'], dtype='M8[ns]')
+        expected = np.array(['2015-01-01T00:00:00.000000000+0000',
+                             '2015-01-03T00:00:00.000000000+0000'], dtype='M8[ns]')
+        result = algos.unique(arr)
+        result.sort()
+        tm.assert_numpy_array_equal(result, expected)
```

Comment:
construct the expected in the correct order. The point is to guarantee that the ``.unique`` ordering is the seen order.